Return empty CSourceModule when no lowered_funcs exists in Relay mod

kumasento opened the pull request:
This PR implements the dummy function idea as mentioned in #4748 - when the whole Relay module is optimized to empty, we can insert a dummy operator that allows TVM to still produce a library.

tqchen commented:
cc @zhiics @FrozenGene

mbarrett97 commented on src/relay/backend/build_module.cc:
```diff
@@ -438,13 +442,19 @@ class RelayBuildModule : public runtime::ModuleNode {
+      Stmt body = EvaluateNode::make(0);
+      Array<ObjectRef> api_args;
+      auto dummy_func = MakeAPI(body, "__dummy__", api_args, 0, false);
+      lowered_funcs.Set("llvm", Array<LoweredFunc>({dummy_func}));
```
Is defaulting the LLVM the correct behaviour here (eg. will this fall over if we build without LLVM support)?

FrozenGene replied:
I think should set `target_host_`. Even we have LLVM support, it is not correct too, imagine our target host is ARM.

kumasento replied:
Thank you guys for your kind comments. We don't need to set target in the latest commit.

mbarrett97 commented on src/relay/backend/build_module.cc:
```diff
@@ -438,13 +442,19 @@ class RelayBuildModule : public runtime::ModuleNode {
-        lowered_funcs,
-        target_host_,
-        BuildConfig::Current());
+      LOG(WARNING) << "No lowered funcs exist in the compiled module, "
```
Do we need to retain this warning? With external codegen, having no lowered funcs can be a perfectly normal mode of operation.

kumasento replied:
Thanks @mbarrett97 , I've removed that log.

tqchen commented:
I agree that perhaps an empty module provides useful middle ground. The closest thing so far might be CSourceModule with an empty string https://github.com/apache/incubator-tvm/blob/master/src/target/source/source_module.cc#L190

zhiics commented:
CSourceModule with an empty string looks to me as well. @kumasento could you do that instead of creating a dummy llvm module? Thanks.

kumasento commented:
Thank you guys for all your kind reviews! @mbarrett97 @FrozenGene @tqchen @zhiics 

I've updated this PR to fulfill the following revisions:

1. Using `CSourceModule` with an empty string as the returned module object `ret_.mod`, instead of creating a dummy lowered function.
2. Removed the warning based on @mbarrett97 review.

Now the generated module looks clean and tidy. No redundant dummy function generated and no extra design decisions should be made. 

Please let me know if there is anything else you feel should be done. Thanks!

zhiics commented on src/relay/backend/build_module.cc:
```diff
@@ -28,15 +28,20 @@
 #include <tvm/relay/expr.h>
 #include <tvm/relay/transform.h>
 #include <tvm/relay/qnn/transform.h>
+#include <tvm/tir/ir_pass.h>
```
remove?

zhiics commented on src/relay/backend/build_module.cc:
```diff
@@ -438,13 +439,14 @@ class RelayBuildModule : public runtime::ModuleNode {
     auto lowered_funcs = graph_codegen_->GetLoweredFunc();
     if (lowered_funcs.size() == 0) {
-      LOG(WARNING) << "no lowered funcs exist in the compiled module";
+      ret_.mod = tvm::codegen::CSourceModuleCreate(std::string(), std::string());
```
just `codegen::CSourceModuleCreate("", "")` ?

zhiics commented on src/relay/backend/build_module.cc:
```diff
@@ -438,13 +439,14 @@ class RelayBuildModule : public runtime::ModuleNode {
     auto lowered_funcs = graph_codegen_->GetLoweredFunc();
     if (lowered_funcs.size() == 0) {
-      LOG(WARNING) << "no lowered funcs exist in the compiled module";
+      ret_.mod = tvm::codegen::CSourceModuleCreate("", "");
```
Sorry for the back and forth. Could you please add a comment here so that ppl would know what we are doing here?

zhiics replied:
You can force push so that your previous CI could be terminated earlier.

kumasento replied:
Sure thing, just added that

mbarrett97 commented:
I don't think the empty CSourceModule method works. There's a check in source_module.cc that fails when you try and create one with an empty string.

kumasento commented:
> I don't think the empty CSourceModule method works. There's a check in source_module.cc that fails when you try and create one with an empty string.

Hi @mbarrett97 thanks for your comment. Currently I haven't met such an issue while testing. Would you mind letting me know which assertion you were referring to?

kumasento commented:
@mbarrett97 Thanks, I just noticed that the base of my PR is not the latest commit. I will update it soon.

kumasento commented:
Hi @mbarrett97 

Sorry I have got some questions for the case you've mentioned. 

1. As you suggested, the assertion for `lowered_funcs.size()` in the branch assuming that external modules exist should be raised in the current implementation, but what is the expected behaviour when we actually use `CSourceModule` with an empty string? Should we simply remove this assertion since `ret_.mod` exists anyway and won't affect external module importing?
2. This line of code:

https://github.com/apache/incubator-tvm/blob/a5661611472c8e92b20bbe4d074333b8183f2878/src/target/source/source_module.cc#L101

is located in a `SaveToFile` function, which is not called frequently, and the single use case I've noticed:

https://github.com/apache/incubator-tvm/blob/fc1a1d83691f7229e25a37befdcd3ad9c3d430b3/tests/cpp/utvm_runtime_standalone_test.cc#L100

only exists in test and doesn't trigger that assertsion since `fmt` is `"o"`. Even if `fmt` is `"cc"`, assuming no such assertion and the string is empty, there should not be any exception caused (at the code level, I'm not sure whether functionally it is expected). So do you think we could simply remove that empty string checker? 

Thanks!

mbarrett97 commented:
I should probably explain my 'test case' :) Apologies if you know this all already. I'm using the external codegen feature which is relatively new. Essentially it allows you to take parts of the Relay graph and run them through a non-TVM compiler. In some cases, we take the entire Relay graph and put it through one of these external compilers resulting in the case where there are no lowered TVM functions. What does get produced though is 'external' runtime modules which correspond to the outputs from these external compilers.

You can see this happening in the BuildRelay function. We call 'GetExternalModules' to get those modules that have been produced by external compilers. But to be able to compile these modules into a shared library, we need a 'host' TVM module. This host module then 'imports' in all the external modules, as you can see here:
https://github.com/apache/incubator-tvm/blob/a5661611472c8e92b20bbe4d074333b8183f2878/src/relay/backend/build_module.cc#L458

However, if there is no host TVM module (which will happen when there are no lowered funcs), then you can't import the external modules into a host module and accordingly we fail to compile the result to a shared library. So this fix is very helpful by making sure that host module always gets created, even if it's empty, so we can always compile to a shared library.

Now addressing your questions. For 1, yes I think we just remove this branch entirely and have it always import the external modules. For 2, I hit the assert when trying to call export_library on the 'lib' output of the build function. I believe SaveToFile is called when doing this export, in particular this line:
https://github.com/apache/incubator-tvm/blob/79cfab00d1a5ae55b3324dae803d5e1389e12bbe/python/tvm/runtime/module.py#L302
I did try just commenting out the assert and it seems to work. However, I then ran into a new problem... Because what you produce is ostensibly a 'c source' file (even if it's empty), all the external modules are also exported as c files. This results in having to do a proper compile when you try and merge all the modules together to produce a shared library. If one of your external modules happens to be a binary artifact that's 150MB (my case), this compile uses an incredible amount of RAM and takes a long time to complete. I didn't get this problem with your first solution as it used an LLVMModule (I think?) which results in the external modules becoming object files so all you need to do is link, not perform a full compilation.

Sorry for the wall of text, if you have any further questions I'll happily answer them. I realise I may not have perfectly explained everything here :)

kumasento commented:
@mbarrett97 Thank you for your detailed explanation! I kind of understand what's going here.

For Q1, I have updated the logic involving external module importing. It now always import external modules into `ret_.mod` since it is always there.

For Q2, I'm afraid that I don't have a better idea for that. Maybe @tqchen @zhiics @FrozenGene can have an input on this? Thanks!

tqchen commented:
The reason that LLVMModule simplifies compilation is because it remembers the correct target triple. We can try to enhance the code by introducing an Empty LLVM module(with the correct target) when it is available(so that Q2 can be resolved) and fallback to the CSourceModule.

zhiics commented:
For Q2, a simpler would be feeding it with ";" instead of empty string. But LLVM module should be faster for compilation.

FrozenGene commented:
> I did try just commenting out the assert and it seems to work. However, I then ran into a new problem... Because what you produce is ostensibly a 'c source' file (even if it's empty), all the external modules are also exported as c files. This results in having to do a proper compile when you try and merge all the modules together to produce a shared library. If one of your external modules happens to be a binary artifact that's 150MB (my case), this compile uses an incredible amount of RAM and takes a long time to complete. I didn't get this problem with your first solution as it used an LLVMModule (I think?) which results in the external modules becoming object files so all you need to do is link, not perform a full compilation.

This is because our new `ModulePackImportsToLLVM` simplifies complication and generate object file directly, it is designed for large artifact like your case.

> The reason that LLVMModule simplifies compilation is because it remembers the correct target triple. We can try to enhance the code by introducing an Empty LLVM module(with the correct target) when it is available(so that Q2 can be resolved) and fallback to the CSourceModule.
  cc @FrozenGene who created the mechanism for the binary exporting

I agree we could create empty llvm module if we have llvm target support, then we could fallback to CSourceModule.  We could refer logic here:
```cpp
runtime::Module build(const Map<Target, Array<LoweredFunc>>& inputs,
                      const Target& target_host,
                      const BuildConfig& config)
  Target target_host_val = target_host;
  if (!target_host.defined()) {
    for (const auto& it : inputs) {
      if (it.first->device_type == kDLCPU) {
        target_host_val = it.first;
        break;
      }
    }
  }

  if (!target_host_val.defined()) {
    target_host_val = DefaultTargetHost(target_host_val);
  }

```
Here, we have logic to detect and extract the correct target host. According to this, we could know whether we have LLVM (with correct target).

kumasento commented:
Thank you for your valuable suggestions @tqchen @zhiics @FrozenGene !

I now changed the logic to try `LLVMModule`, and if the `target_host` is not `llvm`, we fallback to `CSourceModule`. I use the code snippet that @FrozenGene suggested to decide the target host. But I still need to bring back the dummy function idea, otherwise, I cannot find a way to generate a `LLVMModule`, which requires `lowered_funcs` to be non-empty.

@mbarrett97 Please see whether this version suits your scenario. Any other suggestions are also welcomed!

FrozenGene commented on src/relay/backend/build_module.cc:
```diff
@@ -437,28 +441,50 @@ class RelayBuildModule : public runtime::ModuleNode {
+      if (!target_host_val.defined())
+        target_host_val = (pf != nullptr) ? target::llvm() : target::stackvm();
+
+      if (target_host_val.defined() && target_host_val->str() == "llvm")
```
Please add one test "llvm -device=arm_cpu -mattr=+neon -target=aarch64-linux-gnu" to make sure we get correct llvm target. I doubt your `target_host_val->str() == "llvm"` is not correct logic.

kumasento replied:
Thanks @FrozenGene, this is indeed problematic. I changed the condition to `target_host_val->target_name == "llvm"`. Hope that's OK.

mbaret commented:
I've tested the most recent commit with my use-case (aarch64 with offloading to external codegen) and it looks to work well. It's still not a particularly elegant solution but I can't propose a better one.

FrozenGene commented on src/relay/backend/build_module.cc:
```diff
@@ -437,28 +441,51 @@ class RelayBuildModule : public runtime::ModuleNode {
+        lowered_funcs.Set(
+          target_host_val->str(),
+          Array<LoweredFunc>({
+             MakeAPI(EvaluateNode::make(0), "__dummy__", Array<ObjectRef>(), 0, false) }));
```
Add one `__dummy__` is not elegant solution IMO. If we have this LLVM IR:
```
target datalayout = "e-m:o-i64:64-f80:128-n8:16:32:64-S128"
target triple = "x86_64-apple-macosx10.14.0"
```
Does this satisfy your need?

If it is, I think we could create one packed function in `llvm_module.cc`. We could create this kind of empty llvm module. Then we could have`ret_.mod = tvm::codegen::LLVMModuleCreate...` like CSourceModuleCreate.

kumasento replied:
This sounds like a better way to go. I can try this out later.

tqchen commented:
@kumasento @FrozenGene can you followup on this PR?

FrozenGene commented:
> @kumasento @FrozenGene can you followup on this PR?

Sure. I am waiting @kumasento 's new implementation as comments said.

kumasento commented:
Yes, I will do that very soon, was trapped at a conference. Sorry for the
delay.

On Wed, 26 Feb 2020 at 05:00, Zhao Wu <notifications@github.com> wrote:

> @kumasento <https://github.com/kumasento> @FrozenGene
> <https://github.com/FrozenGene> can you followup on this PR?
>
> Sure. I am waiting @kumasento <https://github.com/kumasento> 's new
> implementation as comments said.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/apache/incubator-tvm/pull/4847?email_source=notifications&email_token=ACC42R7WHOV6YFUXC32V4QDREXZPNA5CNFSM4KR43YZ2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEM6ZTZA#issuecomment-591239652>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ACC42R6SLTCAU4RARSUMEOLREXZPNANCNFSM4KR43YZQ>
> .
>

kumasento commented:
@FrozenGene @tqchen @zhiics @mbaret Would you mind giving another read on this? Now the approach is creating an empty LLVM module instead of using a dummy function. There is also a `LLVMModuleCreate` packed function implemented.

FrozenGene commented on src/target/llvm/llvm_module.cc:
```diff
@@ -356,6 +356,40 @@ TVM_REGISTER_GLOBAL("codegen.build_llvm")
+  auto ctx = std::make_shared<llvm::LLVMContext>();
+  llvm::SMDiagnostic err;
+  auto mem_buf = llvm::MemoryBuffer::getMemBuffer(ir_str);
+  auto module = llvm::parseIR(mem_buf->getMemBufferRef(), err, *ctx);
```
I think it is not one elegant solution to create LLVM module using ir string concatenation. You could use the api like this:
```cpp
  auto triple = tm->getTargetTriple();
  auto ctx = std::make_shared<llvm::LLVMContext>();
  std::string module_name = "dummy";
  std::unique_ptr<llvm::Module> module(new llvm::Module(module_name, *ctx));
  module->setTargetTriple(triple.str());
  module->setDataLayout(tm->createDataLayout());
```
Does this work well too? You could refer more detail in my previous implementation of codegen_blob.cc

kumasento replied:
@FrozenGene Thank you very much! It looks much better. I've updated my PR accordingly.

tqchen commented:
ping @FrozenGene please followup

FrozenGene commented on src/target/llvm/llvm_module.cc:
```diff
@@ -356,6 +356,26 @@ TVM_REGISTER_GLOBAL("codegen.build_llvm")
+  auto target = args[0].operator std::string();
+  auto module_name = args[1].operator std::string();
+
+  // create a default data layout
```
Sorry for later response. Minor comment: The logic here doesn't only create default data layout but also create default target triple. We should update the comment.

kumasento replied:
Thanks @FrozenGene I've updated the comments based on your comments :)

FrozenGene commented:
@mbaret @zhiics @tqchen could take a look another round.

tqchen commented:
Thanks @kumasento @FrozenGene @mbaret @zhiics  This PR is now merged

trevor-m commented:
This commit is causing segfaults when my entire relay program is offloaded to an external codegen.

kumasento commented:
Hi Trevor,

Thanks for the info. Can you provide more information about the bug you've
got? Thanks.


On Tue, 17 Mar 2020 at 17:29, Trevor Morris <notifications@github.com>
wrote:

> This commit is causing segfaults when my entire relay program is offloaded
> to an external codegen.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/apache/incubator-tvm/pull/4847#issuecomment-600200202>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ACC42R5SF23UYPEF6M5ABT3RH6XPNANCNFSM4KR43YZQ>
> .
>

trevor-m commented:
HI @kumasento the graph runtime is trying to get my external codegen functions from the empty LLVM module.

Stack trace shows the segfault is coming from LLVMModuleNode, so I added a print statement after [this line](https://github.com/apache/incubator-tvm/blob/master/src/target/llvm/llvm_module.cc#L59) which showed that GetFunction() was called on the LLVM module for my function `tensorrt_0` which should be going to my external codegen module instead.

Stack trace:
```
  [bt] (0) /home/ubuntu/anaconda3/lib/python3.6/site-packages/mxnet/libmxnet.so(+0x2e6b140) [0x7f3b510c0140]
  [bt] (1) /lib/x86_64-linux-gnu/libc.so.6(+0x354b0) [0x7f3bdd7134b0]
  [bt] (2) /lib/x86_64-linux-gnu/libc.so.6(strlen+0x26) [0x7f3bdd769746]
  [bt] (3) /data/neo-ai-tvm/build/libtvm.so(tvm::codegen::LLVMModuleNode::LazyInitJIT()+0x894) [0x7f3bcca32ad4]
  [bt] (4) /data/neo-ai-tvm/build/libtvm.so(tvm::codegen::LLVMModuleNode::GetFunction(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, tvm::runtime::ObjectPtr<tvm::runtime::Object> const&)+0x488) [0x7f3bcca334b8]
  [bt] (5) /data/neo-ai-tvm/build/libtvm.so(tvm::runtime::ModuleNode::GetFunction(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, bool)+0x45) [0x7f3bcca757f5]
  [bt] (6) /data/neo-ai-tvm/build/libtvm.so(tvm::runtime::GraphRuntime::CreateTVMOp(tvm::runtime::TVMOpParam const&, std::vector<DLTensor, std::allocator<DLTensor> > const&, unsigned long)+0x4d5) [0x7f3bccad8e05]
  [bt] (7) /data/neo-ai-tvm/build/libtvm.so(tvm::runtime::GraphRuntime::SetupOpExecs()+0x661) [0x7f3bccadb0d1]
  [bt] (8) /data/neo-ai-tvm/build/libtvm.so(tvm::runtime::GraphRuntime::Init(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, tvm::runtime::Module, std::vector<DLContext, std::allocator<DLContext> > const&)+0x260) [0x7f3bccadcf60]
```

kumasento commented:
Thanks @trevor-m 

@FrozenGene sorry for bothering you but does this ring a bell?

I feel it is weird that `GetFunction` won't return an invalid/empty value directly when the function name cannot be found. I'm wondering why `LazyJIT` should be called in this scenario?

FrozenGene commented:
> Thanks @trevor-m
> 
> @FrozenGene sorry for bothering you but does this ring a bell?
> 
> I feel it is weird that `GetFunction` won't return an invalid/empty value directly when the function name cannot be found. I'm wondering why `LazyJIT` should be called in this scenario?

Please see: https://github.com/apache/incubator-tvm/pull/4847/files#diff-8baddb83a9684e8373691bb48a946900R469-R474 When entire program is offloaded, I find previous logic will 
```cpp
// Execute the whole module using external runtime.
        ret_.mod = ext_mods[0];
```

However, current logic we will
```cpp
// Import all external runtime modules.
    for (const auto& it : ext_mods)
      ret_.mod.Import(it);
```

Could you double check the logic is the same as previous? Thanks.

kumasento commented:
Hi @FrozenGene 

Thanks for your explanation. I do have a question about this part, hope you won't mind:

1. Before this PR, we replace `ret_.mod` with `ext_mods[0]` when there is no `lowered_funcs`, which is sensible since no `ret_.mod` is available when `lowered_funcs` does not exist.
2. Now we will always create a `ret_.mod`. What would the logic be if there is only one external module (the condition for `ret_.mod` replacement)? Should we do replacement or import? 

Also @trevor-m would you mind sending me a minimal workable example? I would like to do the tracing myself as well.

Thanks

FrozenGene commented:
> Hi @FrozenGene
> 
> Thanks for your explanation. I do have a question about this part, hope you won't mind:
> 
> 1. Before this PR, we replace `ret_.mod` with `ext_mods[0]` when there is no `lowered_funcs`, which is sensible since no `ret_.mod` is available when `lowered_funcs` does not exist.
> 2. Now we will always create a `ret_.mod`. What would the logic be if there is only one external module (the condition for `ret_.mod` replacement)? Should we do replacement or import?
> 
> Also @trevor-m would you mind sending me a minimal workable example? I would like to do the tracing myself as well.
> 
> Thanks

I think we should be `ret_.mod` replacement as previous pr. @zhiics do the external codegen part. He could answer it more authority. @zhiics Could you help to answer this question?

zhiics commented:
I think changing it to a llvm module and import all submodules is okay. Now if you only have an external module. You will need to create a llvm module first and them import the external module to it. 

Stepping into llvm module to find the symbol is not wrong because we will always try to find the symbol from the host module first. If it is not found, we will then try to check each imported module. See the code here:

https://github.com/apache/incubator-tvm/blob/050f2bde2c694af9b5569ca954ca041c3767787b/src/runtime/module.cc#L65

A minimal example to reproduce this and track the root cause would be more helpful.

trevor-m commented:
> I think changing it to a llvm module and import all submodules is okay. Now if you only have an external module. You will need to create a llvm module first and them import the external module to it.
> 
> Stepping into llvm module to find the symbol is not wrong because we will always try to find the symbol from the host module first. If it is not found, we will then try to check each imported module. See the code here:
> 
> https://github.com/apache/incubator-tvm/blob/050f2bde2c694af9b5569ca954ca041c3767787b/src/runtime/module.cc#L65
> 
> A minimal example to reproduce this and track the root cause would be more helpful.

You can reproduce this by running `test_extern_dnnl()` after commenting out this line: https://github.com/apache/incubator-tvm/blob/master/tests/python/relay/test_pass_partition_graph.py#L203

kumasento commented:
Hi @trevor-m 
Thanks for this information. I ran the test and reproduced the bug. I've located that the segfault should be raised from the `create` function of `llvm::EngineBuilder` (this [line](https://github.com/apache/incubator-tvm/blob/master/src/target/llvm/llvm_module.cc#L316)).

Now I'm looking at the internal logic in LLVM to find out what the actual cause is. Please bear with me for 1-2 days.

kumasento commented:
@trevor-m a tentative fix has been posted in #5146
